pulsar-broker/src/test/java/org/apache/pulsar/client/api/MutualAuthenticationTest.java

```diff
@@ -0,0 +1,225 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+import static org.testng.Assert.assertEquals;
+
+/**
+ * Test Mutual Authentication.
+ * Test connect set success, and producer consumer works well.
+ */
+public class MutualAuthenticationTest extends ProducerConsumerBase {
+    private static final Logger log = LoggerFactory.getLogger(MutualAuthenticationTest.class);
+
+    private static String[] clientAuthStrings = {
+        "MutualClientAuthInit", // step 0
+        "MutualClientStep1"     // step 1
+    };
+
+    private static String[] serverAuthStrings = {
+        "ResponseMutualClientAuthInit", // step 0
+        "ResponseMutualClientStep1"     // step 1
+    };
+
+    private static int authStep = 0;
+
+    public static class MutualAuthenticationDataProvider implements AuthenticationDataProvider {
+        @Override
+        public boolean hasDataFromCommand() {
+            return true;
+        }
+
+        @Override
+        public AuthData authenticate(AuthData data) throws IOException {
+            String dataString = new String(data.getBytes(), Charset.forName("UTF-8"));
+            if (!dataString.equalsIgnoreCase("init")) {
+                assertEquals(dataString, serverAuthStrings[authStep - 1]);
+            }
+            log.debug("authenticate in client. passed in :{}, send: clientAuthStrings[{}]: {}",
+                dataString, authStep, clientAuthStrings[authStep]);
+            return AuthData.of(clientAuthStrings[authStep].getBytes());
+        }
+    }
+
+    public static class MutualAuthentication implements Authentication {
+        @Override
+        public void close() throws IOException {
+            // noop
+        }
+
+        @Override
+        public String getAuthMethodName() {
+            return "MutualAuthentication";
+        }
+
+        @Override
+        public AuthenticationDataProvider getAuthData(String broker) throws PulsarClientException {
+            try {
+                return new MutualAuthenticationDataProvider();
+            } catch (Exception e) {
+                throw new PulsarClientException(e);
+            }
+        }
+
+        @Override
+        public void configure(Map<String, String> authParams) {
+            // noop
+        }
+
+        @Override
+        public void start() throws PulsarClientException {
+            // noop
+        }
+    }
+
+
+    public static class MutualAuthenticationState implements AuthenticationState {
+        @Override
+        public String getAuthRole() throws AuthenticationException {
+            return "admin";
+        }
+
+        @Override
+        public AuthData authenticate(AuthData authData) throws IOException {
+            String clientData = new String(authData.getBytes(), Charset.forName("UTF-8"));
+            assertEquals(clientData, clientAuthStrings[authStep]);
```

Comment:
Don't assert like this.  This assertion isn't happening in the main testng thread, so it will never report correctly. For the server side stuff, any failure should result in AuthenticationException.